wall/wall.c

```diff
@@ -0,0 +1,125 @@
+/*
+ * wall.c - write to all users
+ *
+ * Copyright (C) 2024: Luiz Antônio Rangel (takusuman)
+ *
+ * SPDX-Licence-Identifier: Zlib
+ *
+ *	from Unix 32V /usr/src/cmd/wall.c
+ *	December 19th, 1978.
+ *
+ * Copyright(C) Caldera International Inc. 2001-2002. All rights reserved.
+ *
+ * SPDX-Licence-Identifier: Caldera
+ */
+
+#include <errno.h>
+#include <limits.h>
+#include <pwd.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strmenta.h>
+#include <sys/types.h>
+#include <unistd.h>
+#include <utmp.h>
+#include <utmpx.h>
+#define MAXNAMLEN _POSIX_LOGIN_NAME_MAX
+
+char	*mesg = "",
+	who[MAXNAMLEN] = "???",
+	sterm[32] = "";
+int	msize = 0;
+void main(int argc, char *argv[]);
+void sendmes(char *tty);
+
+void main(int argc, char *argv[]) {
+	int i = 0;
+	char c = '\0',
+	     smesg[3000] = "";
+	struct utmpx *utmp;
+	struct passwd *pw;
+	FILE *f = NULL;
+
+	if ((utmp = getutxent()) == NULL) {
+		fprintf(stderr, "failed to open utmpx database: %s\n",
+			strerror(errno));
+		exit(1);
+	}
+
+	f = stdin;
+	if(argc >= 2) {
+		/* take message from unix file instead of standard input */
+		if((f = fopen(argv[1], "r")) == NULL) {
+			fprintf(stderr, "Cannot open %s\n", argv[1]);
+			exit(1);
+		}
+	}
+	
+	for (; ((i = getc(f)) != EOF); ) smesg[msize++] = i;
+	fclose(f);
+	mesg = ssafe(smesg);
+	
+	pw = getpwuid(geteuid());
+	if (pw) {
+		for (i = 0; c = pw->pw_name[i]; i++) who[i]=c;
+		who[i] = '\0'; /* sender initials */
+	}
+
+	/* Get the current terminal */
+	strncpy(sterm, ttyname(fileno(stderr)), sizeof(sterm));
+	for (i = 1; sterm[i] != '/'; i++);
+	strncpy(sterm, &sterm[(i + 1)], UT_LINESIZE);
+
+	/* Rewind utmpx file to its start,
+	 * like a cassete tape. */
+	setutxent();
+	for (; utmp = getutxent(); ) {
+		switch (utmp->ut_type) {
+			case USER_PROCESS:
+				if (utmp != NULL) {
+					sleep(1);
+					sendmes(utmp->ut_line);
+				} else {
+					break;
+				}
+			default:
+				continue;
+		}
+		break;
```

Comment:
This break is unreachable. The
```c
		default:
				continue;
		}
		break;
```
could be replaced by
```c
		}
```

Reply:
Actually, I think it might be necessary because of the conditional above that checks it the utmp struct isn't empty.

Reply:
On a second thought, maybe checking if utmp is not NULL would have to take preference over checking if ``utmp->ut_type`` is ``USER_PROCESS``. I will do this tomorrow, it's too late right now.

Reply:
> Actually, I think it might be necessary because of the conditional above that checks it the utmp struct isn't empty.

You're right. I forgor the break doesnt break out of the loop

Reply:
If you don't pretend to add many other checks soon, I believe an `if` would be cleaner

Reply:
> If you don't pretend to add many other checks soon, I believe an `if` would be cleaner

I mean, it is indeed faster to use a switch.

Reply:
> > If you don't pretend to add many other checks soon, I believe an `if` would be cleaner
> 
> I mean, it is indeed faster to use a switch.

Why?

Reply:
> > > If you don't pretend to add many other checks soon, I believe an `if` would be cleaner
> > 
> > 
> > I mean, it is indeed faster to use a switch.
> 
> Why?

On C --- I don't know about Go or even C++ in this case --- the assembly code for an if conditional works with ``jump``ing to another segment of the code for comparing, if not, it continues ``jump``ing and comparing until it comes to the end of the condition. But a switch case works differently, similarly to a hashmap.
That's why it is more commonly used on ad-hoc parsers and loops with constant conditionals rather than if, because it is, indeed, faster.
When this ``wall``(1M) implementation was created, you didn't had to worry with multiple users logged on a machine since not even Xorg existed at the time, but nowadays you can have a user logged 3, 5, even 10 times depending on how much terminals he/she has opened on the environment, that's why using a switch here may have some difference --- it also would be a matter of correctness rather than performance, in my opinion.